Update versions, build on GitHub Actions

camscale opened the pull request:
Update the Centos 8 Dockerfile so that it works as the yum archive has
been moved to `vault.centos.org`.

Add a new Debian 12 Dockerfile as the Debian 8 one no longer builds due
to the signing key of that Debian release having expired.

Pin fpm version in the Dockerfiles to 1.15.1.

Add a GitHub Actions workflow to build and push the centos8 and debian12
images, and remove the `.drone.yml` Drone config.

Issue: https://github.com/gravitational/teleport/issues/20729

camscale committed:
Update Centos 8 Dockerfile to be buildable

Update the Centos 8 Dockerfile so that it builds. Package archives for
Centos 8 have been moved to `vault.centos.org`, so update the yum config
to reflect this.

Also pin the version of fpm to the current latest - 1.15.1.

camscale committed:
Add Debian 12 Dockerfile

Add a Dockerfile to run fpm on Debian 12. The most recent Debian
Dockerfile in this repository is Debian 8 and that no longer builds as
the signing key for that Debian release has expired.

Also pin the version of fpm to the current latest - 1.15.1.

camscale committed:
ci: Move from Drone to GitHub Actions for CI

Move the building and publishing of the OCI images in this repository
from Drone to GitHub Actions. We are turning down Drone.

wadells commented on .github/workflows/build.yaml:
```diff
@@ -8,9 +8,62 @@
+          role-session-name: "gha-fpm-build=${{ github.run_number }}"
+          role-duration-seconds: 900
+      - name: Log into ECR
+        uses: docker/login-action@v3
```
Can we pin this too?  I've been moving everything but `actions/`, `github/`, `gravitational/` and `./` (local repo) references to pinning.

camscale replied:
Done

wadells commented on .github/workflows/build.yaml:
```diff
@@ -8,9 +8,62 @@
+      - name: Checkout repository
+        uses: actions/checkout@v4
+      - name: Build FPM Image
+        uses: docker/build-push-action@v3
```
pin, please

camscale replied:
Done, and updated to v5!

wadells commented on debian/12/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+ENV DEBIAN_FRONTEND noninteractive
+
+RUN set -x \
+	&& apt-get update && apt-get install -y --no-install-recommends \
+		ruby \
+		ruby-dev \
+		gcc \
+		make \
+		ca-certificates \
+		libffi-dev \
+		ruby-ffi \
```
Lets clean up after the apt-get (similar to the `yum clean all` above)

https://docs.docker.com/develop/develop-images/instructions/#apt-get

camscale replied:
Done

wadells commented on debian/12/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM debian:12
```
Would it make sense to use the `12-slim` tag here, or is that too much of a change?

camscale replied:
I've done this. I'll do a test build and ensure it still works

wadells commented on .github/workflows/build.yaml:
```diff
@@ -8,9 +8,62 @@
+          VERSION: ${{ inputs.version }}
+          REF_NAME: ${{ github.ref_name }}
+        run: |
+          V="${VERSION:-${REF_NAME}}"
```
I'm not sure `REF_NAME` is safe.  I think git allows stuff like `/` or `+` that docker registries will choke on.

Probably not a problem, as I don't expect these to show up in tag names even if they're technically legal.

camscale replied:
I've fixed this to replace all invalid chars with hyphens.

wadells reviewed (approved):
I think the overall approach here looks good.  CentOS 8 as our RPM publishing platform makes sense given the options available.

camscale committed:
Pin docker actions in GitHub Actions workflow

camscale committed:
Clean up after apt-get for smaller/cleaner image

camscale committed:
Use debian:12-slim as a base image.

camscale committed:
Sanitize version/ref_name for use as docker label

Docker labels/tags can contain only alphanumeric chars, underscores,
periods and hyphens and must start with alphanumeric or underscores. So
we convert all invalid chars to hyphens and strip leading invalid chars.
Also collapse runs of hyphens to a single hyphen.

Technically we do not need to strip the leading chars as we construct
the label with the version as a suffix, but it would look a little weird
if we didn't and allows us to use the version as a pure tag if we want
to later.

camscale commented:
Running a test of the release-linux workflow that uses the latest images I've built with the current commit in this PR. Should be done shortly, which will verify if debian slim works fine.

camscale commented:
Works fine with debian slim.
